scanner/scanner.go

```diff
@@ -0,0 +1,150 @@
+// instructs the caller to sleep after this call when there are no records to
+// be scanned.
+func (s Scanner) SweepAndScan(abort chan bool) (sleepAfter bool) {
+	sl, err := s.staticDB.SweepAndLock(s.staticCtx)
+	if err != nil && !errors.Contains(err, database.ErrNoDocumentsFound) {
+		s.staticLogger.Warnf("error while trying to lock a new record: %s", err)
+		return
+	}
+	if errors.Contains(err, database.ErrNoDocumentsFound) {
+		// Sleep after this run.
+		return true
+	}
+	if sl.Skylink == "" {
+		s.staticLogger.Warnf("empty skylink in record %+v", sl)
+		build.Critical(fmt.Sprintf("SweepAndLock returned a record with an empty skylink. Record hash: %s", hex.EncodeToString(sl.Hash[:])))
+		return
+	}
+	inf, desc, size, scannedSize, err := s.staticClam.ScanSkylink(sl.Skylink, abort)
+	if err != nil {
+		// Scanning failed, log the error and unlock the record for another attempt.
+		s.staticLogger.Debugln(errors.AddContext(err, "scanning failed"))
+		sl.Status = database.SkylinkStatusNew
+		sl.Timestamp = time.Now().UTC()
+		err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+		if err != nil {
+			s.staticLogger.Debugln(errors.AddContext(err, "unlocking a skylink failed"))
+		}
+		return
+	}
+	// Sanity check: scannedSize vs size.
+	if scannedSize > size {
+		build.Critical(fmt.Sprintf("Scanned size (%d bytes) is more than the content size (%d bytes) for skylink %s", scannedSize, size, sl.Skylink))
+	}
+	// Once the scanning is complete, we no longer need the skylink. We want to
+	// remove it and only keep its hash. We don't want our database to be an
+	// index of nasty files.
+	sl.Skylink = ""
+	sl.Status = database.SkylinkStatusComplete
+	sl.Infected = inf
+	sl.InfectionDescription = desc
+	sl.Size = size
+	sl.ScannedAllContent = scannedSize == size
+	sl.ScannedAllOffsets = false
+	sl.Timestamp = time.Now().UTC()
+	err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+	if err != nil {
+		s.staticLogger.Debugln(errors.AddContext(err, "updating a skylink's status failed"))
+	}
+	return
+}
+
+// Start launches a background task that periodically scans the database for
+// new skylink records and sends them for scanning.
+func (s Scanner) Start() {
+	// abort channel which interrupts the current scanning operation
+	abort := make(chan bool, 1)
+	// sleepLength defines how long the thread will sleep before scanning the
+	// next skylink. Its value is controlled by SweepAndScan - while we keep
+	// finding files to scan, we'll keep this sleep at zero. Once we run out of
+	// files to scan we'll reset it to its full duration of sleepBetweenScans.
+	sleepLength := sleepBetweenScans
+	go func() {
+		for {
+			select {
+			case <-s.staticCtx.Done():
+				// interrupt the current scan and exit
+				abort <- true
+				return
+			case <-time.After(sleepLength):
+				shouldSleep := s.SweepAndScan(abort)
+				s.staticLogger.Traceln("shouldSleep is ", shouldSleep)
+				if shouldSleep {
+					sleepLength = sleepBetweenScans
+				} else {
+					sleepLength = 0
```

Comment:
So I think we should definitely be handling the sleep decision here. There are numerous error conditions that `SweepAndScan` is quietly handling. Some of them, if not most of them, should probably sleep a tiny bit to prevent this loop from rapid cycling on the same error.

Reply:
Re: sleep decision being made here - agreed.

Re: short sleep after some/all errors: That's an interesting idea. How long of a sleep do you reckon will make sense? 10ms? 1s? After which errors should we do that?

We also might need to take a break after N consecutive errors, in order to avoid error loops?

Reply:
Added increasingly long sleeps on subsequent errors.
We start with 100ms, then 1s, 10s, and 100s. The length is reset on a successful call - either a successful scan or a "no skylinks found to scan" error.